applications: nrf_desktop: Battery measurement documentation (#1211)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/battery_meas.rst

|                |             | ``battery_meas`` | ``battery_level_event`` | :ref:`bas`       |
+----------------+-------------+------------------+-------------------------+------------------+

Configuration
*************

The module uses the Zephyr :ref:`zephyr:adc` driver to measure voltage.
For this reason, you should set the following options:

* :option:`CONFIG_ADC`
* :option:`CONFIG_ADC_ASYNC` - implementation uses asynchronous calls.
* :option:`CONFIG_ADC_0` - implementation uses ``ADC_0``.
* :option:`CONFIG_ADC_NRFX_SAADC` - implementation uses nrfx SAADC driver for nRF52 MCU series.

Set ``CONFIG_DESKTOP_BATTERY_MEAS`` to enable the module.
Define ``DESKTOP_BATTERY_MEAS_POLL_INTERVAL_MS`` - time between subsequent battery measurements (in ms).

Set ``CONFIG_DESKTOP_BATTERY_MEAS_HAS_VOLTAGE_DIVIDER`` option if measurement circuit contains voltage divider made of two resistors.

Values of mentioned resistors (in kOhm) are defined as:

* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_UPPER`` - upper resistor.
* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_LOWER`` - lower resistor.

> The module measures voltage on lower resistor.

Reviewer: voltage over the lower resistor.

Battery voltage is converted to state of charge (SoC) according to lookup table defined in ``battery_def.h`` file located in board-specific directory in the application configuration folder.

You should also define battery parameters (default values will be used if options are left undefined):

* ``CONFIG_DESKTOP_BATTERY_MEAS_MIN_LEVEL`` - battery voltage that corresponds to 0% level (in mV).
* ``CONFIG_DESKTOP_BATTERY_MEAS_MAX_LEVEL`` - battery voltage that corresponds to 100% level (in mV).

> * ``CONFIG_DESKTOP_VOLTAGE_TO_SOC_DELTA`` - difference between adjanced elements in conversion LUT (in mV).

Reviewer: adjacent

Author: done

.. _battery_meas:

Battery measurement
###################

The ``battery_meas`` module periodically measures the battery voltage and sends ``battery_level_event`` informing about the current battery level.

Module Events
*************

+----------------+-------------+------------------+-------------------------+------------------+
| Source Module  | Input Event | This Module      | Output Event            | Sink Module      |
+================+=============+==================+=========================+==================+
|                |             | ``battery_meas`` | ``battery_level_event`` | :ref:`bas`       |
+----------------+-------------+------------------+-------------------------+------------------+

Configuration
*************

The module uses the Zephyr :ref:`zephyr:adc` driver to measure voltage.
For this reason, you should set the following options:

> * :option:`CONFIG_ADC`

Author: It just enables ADC converter in Zephyr, I will add description.

> * :option:`CONFIG_ADC_ASYNC` - implementation uses asynchronous calls.

Author: I meant that implementation in our application uses asynchronous calls and this is ther reason why this option should be enabled. I will change it to : `the module implementation uses`

* :option:`CONFIG_ADC_0` - implementation uses ``ADC_0``.
* :option:`CONFIG_ADC_NRFX_SAADC` - implementation uses nrfx SAADC driver for nRF52 MCU series.

Set ``CONFIG_DESKTOP_BATTERY_MEAS`` to enable the module.
Define ``DESKTOP_BATTERY_MEAS_POLL_INTERVAL_MS`` - time between subsequent battery measurements (in ms).

Set ``CONFIG_DESKTOP_BATTERY_MEAS_HAS_VOLTAGE_DIVIDER`` option if measurement circuit contains voltage divider made of two resistors.

> Values of mentioned resistors (in kOhm) are defined as:

Author: ok

* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_UPPER`` - upper resistor.
* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_LOWER`` - lower resistor.

The module measures the voltage over the lower resistor.

Battery voltage is converted to state of charge (SoC) according to lookup table defined in ``battery_def.h`` file located in board-specific directory in the application configuration folder.

You should also define battery parameters (default values will be used if options are left undefined):

> * ``CONFIG_DESKTOP_BATTERY_MEAS_MIN_LEVEL`` - battery voltage that corresponds to 0% level (in mV).

Reviewer 2: ```suggestion
* ``CONFIG_DESKTOP_BATTERY_MEAS_MIN_LEVEL`` - battery voltage in mV that corresponds to the 0% battery level.
```

> * ``CONFIG_DESKTOP_BATTERY_MEAS_MAX_LEVEL`` - battery voltage that corresponds to 100% level (in mV).

Reviewer 2: ```suggestion
* ``CONFIG_DESKTOP_BATTERY_MEAS_MAX_LEVEL`` - battery voltage in mV that corresponds to the battery 100% level.
```

> * ``CONFIG_DESKTOP_VOLTAGE_TO_SOC_DELTA`` - difference between adjacent elements in conversion LUT (in mV).

Reviewer 2: ```suggestion
* ``CONFIG_DESKTOP_VOLTAGE_TO_SOC_DELTA`` - difference in mV between the adjacent elements in the conversion lookup table.
```

> If a pin is used to enable battery measurement, option ``CONFIG_DESKTOP_BATTERY_MEAS_HAS_ENABLE_PIN`` must be enabled.

Reviewer 2: ```suggestion
If a pin is used to enable the battery measurement, enable the ``CONFIG_DESKTOP_BATTERY_MEAS_HAS_ENABLE_PIN`` option.
```

> Selected pin number must be defined as ``CONFIG_DESKTOP_BATTERY_MEAS_ENABLE_PIN`` (implementation uses ``GPIO0`` port).

Reviewer 2: ```suggestion
The number of the pin used for this purpose must be defined as ``CONFIG_DESKTOP_BATTERY_MEAS_ENABLE_PIN``, because the implementation uses the ``GPIO0`` port.
```

Author: It is more like implementation always uses GPIO0 port, bout you can select the pin. I will reword this.

> Zephyr :ref:`zephyr:gpio` driver is used to control mentioned pin, make sure :option:`CONFIG_GPIO` is set.

Reviewer 2: ```suggestion
Because the Zephyr :ref:`zephyr:gpio` driver is used to control this pin, also set the :option:`CONFIG_GPIO` option.
```

Implementation details
**********************

> The work (:c:type:`struct k_delayed_work`) starts asynchronous voltage measurement and resubmits itself.

Reviewer 2: ```suggestion
The work (:c:type:`struct k_delayed_work`) starts the asynchronous voltage measurement and resubmits itself.
```

> When it is processed next time, it reads measured voltage.

Reviewer 2: ```suggestion
When it is processed the next time, it reads the measured voltage.
```

This read sequence is repeated periodically.
When the system goes to the low-power state, the work is cancelled.

> During first voltage readout, ADC is automatically calibrated to increase measurements accuracy.

Reviewer 2: ```suggestion
During the first voltage readout, the Analog to Digital Converter is automatically calibrated to increase the measurement accuracy.
```

.. _battery_meas:

Battery measurement
###################

The ``battery_meas`` module periodically measures the battery voltage and sends the ``battery_level_event`` event that informs about the current battery level.

Module Events
*************

+----------------+-------------+------------------+-------------------------+------------------+
| Source Module  | Input Event | This Module      | Output Event            | Sink Module      |
+================+=============+==================+=========================+==================+
|                |             | ``battery_meas`` | ``battery_level_event`` | :ref:`bas`       |
+----------------+-------------+------------------+-------------------------+------------------+

Configuration
*************

The module uses the Zephyr :ref:`zephyr:adc` driver to measure voltage.
For this reason, set the following options:

> * :option:`CONFIG_ADC` - the module implementation uses Zephyr ADC driver

Reviewer 2: Add period.

Author: ok

* :option:`CONFIG_ADC_ASYNC` - the module implementation uses asynchronous calls.
* :option:`CONFIG_ADC_0` - the module implementation uses ``ADC_0``.
* :option:`CONFIG_ADC_NRFX_SAADC` - the module implementation uses nrfx SAADC driver for nRF52 MCU series.

Set ``CONFIG_DESKTOP_BATTERY_MEAS`` to enable the module.
Also, make sure to define ``CONFIG_DESKTOP_BATTERY_MEAS_POLL_INTERVAL_MS``, that is the amount of time between the subsequent battery measurements (in ms).

If the measurement circuit contains a voltage divider made of two resistors, set the ``CONFIG_DESKTOP_BATTERY_MEAS_HAS_VOLTAGE_DIVIDER`` option.
The following defines specify the values of these two resistors (in kOhm):

* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_UPPER`` - upper resistor.
* ``CONFIG_DESKTOP_BATTERY_MEAS_VOLTAGE_DIVIDER_LOWER`` - lower resistor.

The module measures the voltage over the lower resistor.

> The battery voltage is converted to the state of charge percentage level units according to the lookup table defined in the ``battery_def.h`` file.

Reviewer 2: ```suggestion
The battery voltage is converted to the state of charge (in percentage level units) according to the lookup table defined in the ``battery_def.h`` file.
```

You can find this file the in board-specific directory in the application configuration folder.

> Remember to also define the following battery parameters (otherwise, the default values will be used):

Reviewer 2: I don't like how this requirement is hidden among explanations. I think we should find a better way to say what must be defned, and do it in one single place, but with explanations next to it... Cannot think of a way right now.

Author: If possible, I try to use bullet list with explainations next to parameters (if descriptions are short and simple).
In case the module has more advanced configuration options, I try to describe the options in groups. The options in every group are describing one subject, for example battery parameters. I think it is more understandable than putting all the options together in one list.
I think we have to assume that person who wants to configure module will read whole configuration section.
